### ISSUE TITLE
"fatal cuda error" accessing texture inside of shader.

### ISSUE THREAD

**Issue by ForceConstant** [owl-project/owl#187]
I am encountering an issue where if I access a texture inside the closesthit program, I get a crash of below error. I am using code similar to the optix7course sample. Whereas I call `owlGeomSetTexture(geom,"texture",textures[mesh.diffuseTextureID]);` and then createTextures(). But then it crashes when I use `tex2D<float4>(sbtData.texture, tc.x, tc.y);`  Any ideas?

```
CUDA call (cudaMemcpyAsync((void*)d_pointer, h_pointer, sizeInBytes, cudaMemcpyHostToDevice, stream)) failed with code 700 (line 78): an illegal memory access was encountered
fatal cuda error
```

**Comment by ingowald** [comment]
Hard to say without more context, but a few thoughts:
- you say you first do owlGeomSetTexture and *then* createTextures() ? you obviously have to create the textures before you can assign them as parameters.
-  the 'cudaMemcpyAsync' is not actually an error from a shader's memory access, but from a memcopy. You might want to add some OWL_CUDA_SYNC_CHECK() after your owlLaunch2D to check if it's an error in the launch that caused the crash, or something else.
-  you don't possibly have that code online to look at?
- there's several texture samples in owl itself - it might be easier to go from these than trying to port stuff over from the opitx7course sample.

**Comment by ForceConstant** [comment]
@ingowald no I can't post the code, but for your first poing, the textures are being loaded/created from tiny_obj_loader.h like the sample. I will try adding the  OWL_CUDA_SYNC_CHECK()

**Comment by ForceConstant** [comment]
@ingowald 

So here is the result of adding a CUDA_SYNC_CHECK right after Launch2D
```

error (/workspace/src/SampleRenderer.cpp: line 270): an illegal memory access was encountered
fatal cuda error
./flsSensorApp(+0x57cea) [0x561b5d179cea]
./flsSensorApp(+0x5707b) [0x561b5d17907b]
./flsSensorApp(+0xb0472) [0x561b5d1d2472]
./flsSensorApp(+0xd7d60) [0x561b5d1f9d60]
./flsSensorApp(+0xd7b85) [0x561b5d1f9b85]
./flsSensorApp(+0xace95) [0x561b5d1cee95]
/usr/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f002023f0b3]
./flsSensorApp(+0x31d9e) [0x561b5d153d9e]

```

**Comment by ingowald** [comment]
Ahhh, OK, i think I have a suspicion what's going on.
Before I voice that, let me ask: when you said you're "using code similar to optix7course's createTextures()" .... do you mean that you still use actual CUDATextureObject related calls on the host? I.e., something like `cudaCreateTextureObject()` etc, and then just type-cast the resulting cudaTextureObject_t to a OWLTexture ?

**Comment by ForceConstant** [comment]
`  
void SampleRenderer::createTextures()
  {
    int numTextures = (int)model->textures.size();

    textures.resize(numTextures);

    printf("Loading %d Textures\n", numTextures);

    // TODO: Load normal map texture also.
    for (int textureID=0;textureID<numTextures;textureID++) {
      auto texture = model->textures[textureID];

      int32_t width  = texture->resolution.x;
      int32_t height = texture->resolution.y;


      this->textures[textureID]
        = owlTexture2DCreate(context,
                             OWL_TEXEL_FORMAT_RGBA8,
                             width,
                             height,
                             texture->pixel,
                             OWL_TEXTURE_LINEAR,
                             OWL_TEXTURE_WRAP);

      printf("Texture #%d , h=%d w=%d handle=%x\n", textureID, width, height, this->textures[textureID]);
    }
  }
`

...
```
if (mesh.diffuseTextureID >= 0) {
        printf("Registering Texture Map\n");
        owlGeomSet1i(geom,"hasTexture",1);
        assert(mesh.diffuseTextureID < (int)textures.size());
        owlGeomSetTexture(geom,"texture",textures[mesh.diffuseTextureID]);
      } else {
        owlGeomSet1i(geom,"hasTexture",0);
      }
```

**Comment by ingowald** [comment]
Hm. So my guess was wrong; you _are_ using it the right way, no fancy typecast that'd break things. This looks correct.... and since it could already create the textures (ie, it didnt' fail in owlTexture2DCreate) we can also be confident that the texel data was valid, etcpp. Interesting.

**Comment by ingowald** [comment]
and you're checking that 'hasTexture' is true before doing the tex2D call? And not doing any `owlTextureDestroy` anywhere?

**Comment by ingowald** [comment]
one thought: check your code where you "declare" the variables of your mesh for copy-n-paste errors. E.g., you might have declares the "texture" with the right name and type, but maybe with the wrong "offsetof" value .... or migth have declared another variable to accidentally use the offset of the texture, so that other value gets written into the texture handle, or similar. OWL _will_ do some sanity-checking of types on the host (ie, it'll complain if you try to assign an int to a texture), but on the device/SBT size it doesn't have enough information to do so, and will just do binary memcpy's into the offsets provided.... and sometimes that can cause bugs where you expect to read oen value, but actually get something else.

Basically, what your snippets indicate is that you do create textures the right way, that they have valid data (otherwise owlTexture2D create would have failed) and assign them the right way (owlGeomSet would have failed)... but for some reason in the SBTdata there's not a valid texture handle, so tex2D<> makes an invalid memory access.

**Comment by ingowald** [comment]
One thing I'd try to do to try and hunt that down is to not actually do the tex2D<> call, but simply print the value of the cudaTextureObject_t on the device, as if it was an int (maybe only for a certain launch index to avoid being overwhelmed by prints). If its a cuda texture object it's probably a pretty low int value (0,1,2, etc; basically a texture ID). Check if that value looks reasonable; if not, maybe it'll give you some clue as to what else might have been written to that location

**Comment by ForceConstant** [comment]
Ok I went through all of your suggestions, and didn't find anything, and yes the cudaTextureObject_t == 1, when it crashes so that looks good. 

But I have some more information, that it has something to do with the model. So I have 2 models, that are basically the same, both just a ground patch, and using the same mtl file, but Bathy.obj crashes, but the Bathy2.obj one loads the texture just fine. Both were exported from Blender, with seemingly the same settings, so it just have something to do with the mesh. The one that doesn't work is even the much smaller one. Any ideas?

[Bathy.zip](https://github.com/owl-project/owl/files/10556984/Bathy.zip)

**Comment by ForceConstant** [comment]
@ingowald small update, so I am getting closer to the issue. Turns out the crash is because texcoord is 0, and comes out of tiny_obj_loader as zero in the broken obj, but good in the working one. Seeing if i can tell why.

**Comment by ForceConstant** [comment]
Ok I found the reason why, Blender in some cases is not exporting the texture_index in the face element section. So it looks like this `f 3//1 41//1 26//1` note the missing number inside the slashes. I don't know why yet, but thanks for the help.

**Comment by ingowald** [comment]
thx for the heads-up; always happy to help.
BTW this "f a//b ..." is perfectly valid OBJ syntax; it just means that this vertex has only position (a) and normal (b), but no texcoords; typically that is if that mesh doesn't have a texture on it. OBJ is weird.... :-/
